### PR TITLE
[Train] Add FullyShardedDataParallel support to TorchTrainer

### PR DESCRIPTION
**Pull request by markrogersjr**
## Why are these changes needed?
As of version 1.11, PyTorch supports automatically sharding large models via `FullyShardedDataParallel`. This change is necessary to take advantage of this new feature.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by amogkam** [comment]
Wow thanks for the contribution @markrogersjr! Will take a closer look later!

File: python/requirements_ml_docker.txt [hunk 1]
```diff
@@ -5,7 +5,7 @@ tblib
 
 # If you make changes to the torch versions, please also make the corresponding changes to `requirements_dl.txt`!
 -f https://download.pytorch.org/whl/torch_stable.html
-torch==1.9.0+cu111
+torch==1.11.0+cu111
```
Comment (amogkam) [inline python/requirements_ml_docker.txt:8]: Unfortunately our dependencies for CI are a bit of a mess right now so upgrading won't work out of the box.

Instead can we do the following:
- Split out the fsdp test to a separate file (`test_torch_fsdp.py`)
- Add this file to `ray/train/tests/BUILD` with the following tags: `tags = ["team:ml", "exclusive", "gpu_only", "torch_1_11"]`
- Create a new test suite like this: https://github.com/ray-project/ray/blob/master/.buildkite/pipeline.gpu.large.yml#L1-L10, except with the following changes:
```
- label: ":tv: :steam_locomotive: Train GPU tests (PyTorch 1.11) "
  conditions: ["RAY_CI_TRAIN_AFFECTED"]
  commands:
    - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/build/upload_build_info.sh; fi }; trap cleanup EXIT
    - PYTHON=3.7 TRAIN_TESTING=1 ./ci/env/install-dependencies.sh
    # Because Python version changed, we need to re-install Ray here
    - rm -rf ./python/ray/thirdparty_files; rm -rf ./python/ray/pickle5_files; ./ci/ci.sh build
    - pip install -Ur torch==1.11.0+cu113 --extra-index-url https://download.pytorch.org/whl/cu113
    - ./ci/env/env_info.sh
    - bazel test --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=gpu,gpu_only,torch_1_11,-ray_air python/ray/train/...
```

**Comment by markrogersjr** [comment]
@amogkam Thanks for the feedback, I tried to incorporate your suggestions as closely as possible. Several of the tests are still not passing. I tried tweaking the PyTorch 1.11 test configuration to no avail. Any suggestions?

File: .buildkite/pipeline.gpu.large.yml [hunk 2]
```diff
@@ -9,6 +9,19 @@
     - ./ci/env/env_info.sh
     - bazel test --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=gpu,gpu_only,-ray_air python/ray/train/...
```
Comment (amogkam) [inline .buildkite/pipeline.gpu.large.yml:10]:
```suggestion
    - bazel test --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=gpu,gpu_only,-ray_air,-torch_1_11 python/ray/train/...
```

File: .buildkite/pipeline.gpu.large.yml [hunk 3]
```diff
@@ -9,6 +9,19 @@
...
 
+- label: ":tv: :steam_locomotive: Train GPU tests (PyTorch 1.11) "
+  conditions: ["RAY_CI_TRAIN_AFFECTED"]
+  commands:
+    - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/build/upload_build_info.sh; fi }; trap cleanup EXIT
+    - PYTHON=3.7 TRAIN_TESTING=1 INSTALL_HOROVOD=1 ./ci/env/install-dependencies.sh
+    # Because Python version changed, we need to re-install Ray here
+    - rm -rf ./python/ray/thirdparty_files; rm -rf ./python/ray/pickle5_files; ./ci/ci.sh build
+    - pip install -Ur ./python/requirements_ml_docker.txt
+    - pip uninstall torch -y
```
Comment (amogkam) [inline .buildkite/pipeline.gpu.large.yml:20]:
```suggestion
    - PYTHON=3.7 TRAIN_TESTING=1 ./ci/env/install-dependencies.sh
    # Because Python version changed, we need to re-install Ray here
    - rm -rf ./python/ray/thirdparty_files; rm -rf ./python/ray/pickle5_files; ./ci/ci.sh build
```

File: .buildkite/pipeline.gpu.large.yml [hunk 4]
```diff
@@ -9,6 +9,19 @@
...
+    - pip install -U torch==1.11.0+cu113 --extra-index-url https://download.pytorch.org/whl/cu113
+    - ./ci/env/env_info.sh
+    - bazel test --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=gpu,gpu_only,torch_1_11,-ray_air python/ray/train/...
```
Comment (amogkam) [inline .buildkite/pipeline.gpu.large.yml:23]:
```suggestion
    - bazel test --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=gpu,gpu_only,+torch_1_11,-ray_air python/ray/train/...
```

**Review by amogkam — COMMENTED** [review]
Thanks @markrogersjr! Left some comments- I think these should work

**Comment by amogkam** [comment]
@markrogersjr can you also merge in latest master?

**Comment by markrogersjr** [comment]
@amogkam I managed to get most tests to pass, looks like the rest could be failing due to flakiness. Please let me know how I can help from here!

**Comment by amogkam** [comment]
Thanks @markrogersjr, this looks great to me! I just pushed some additional changes, primarily for backwards compatibility. But I will make sure to get this merged in!

**Review by amogkam — APPROVED** [review]
Excellent, thanks @markrogersjr!

**Comment by markrogersjr** [comment]
@amogkam thank you as well, nice work!

**Review by richardliaw — APPROVED** [review]
